partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt

```diff
@@ -209,8 +209,20 @@ class UnsupportedTypeCheckException(message: String) : RuntimeException(message)
  */
 sealed class CollectionType : SingleType() {
     abstract val elementType: StaticType
+    abstract val constraints: Set<CollectionConstraint>
+
+    internal fun validateCollectionConstraints(elementType: StaticType, constraints: Set<CollectionConstraint>) {
+        if (elementType !is StructType && constraints.any { it is TupleCollectionConstraint }) {
+            throw UnsupportedTypeConstraint("Only collection of tuples can have tuple constraints")
+        }
+    }
```

Comment:
This very explicit check makes me think that every Constraint, similar to IonSchema's constraint, needs its own `validate()` function -- and every StaticType would have a `validate()` function.

So, an example code snippet:
```kotlin
sealed class CollectionType() : SingleType() {
    public abstract fun validate(): Unit = this.constraints.forEach { validate(this) }
}
```

Then, the definition of TupleCollectionConstraint (or its variants) could be:
```kotlin
data class PrimaryKey(val names: Set<String>) : Constraint {
    override public fun validate(type: StaticType) {
        when (type) {
            is CollectionType -> {
                if (type.elementType !is StructType) {
                    throw SomeException()
                }
            }
            else -> throw SomeException() // Or, for IonSchema, collect Violations
        }
    }
}
```

The PrimaryKey constraint, though, still doesn't make that much sense to me. Because, without external logic, how would we validate a type or ExprValue with a PrimaryKey constraint?

It also makes me think about what the attribute `elementType` for `CollectionType` is. It really is just a constraint at the end of the day. So, for a "TupleCollectionConstraint", why not just, when creating a CollectionType representing some table/relation -- set the elementType to be StructType?

Reply:
> The PrimaryKey constraint, though, still doesn't make that much sense to me. Because, without external logic, how would we validate a type or ExprValue with a PrimaryKey constraint?

Knowing constraints have multiple utilities, one of which is when we want to do optimizations. Consider the following query:

```SQL
SELECT * FROM MyTable AS t WHERE t.id = 10
```

In this case, if we know `id` is a `MyTable`'s primary key we can perform data retrieval from Index rather than full table scan and then filter.

> It also makes me think about what the attribute elementType for CollectionType is. It really is just a constraint at the end of the day. So, for a "TupleCollectionConstraint", why not just, when creating a CollectionType representing some table/relation -- set the elementType to be StructType?

b/c `Collection<StructType>` can be both represent a semi-structured model (E.g. a DynamoDB table) and structured (E.g. a PostgresSQL table). On the other hand, some constraints are only applicable to a collection rather another type i.e. what does it mean to have a `PrimaryKey` or `Check` constraint on a `StructType`?